### DISCUSSION TITLE
pre-compile Tracee's ebpf probe for different kernels

### DISCUSSION THREAD

**Discussion by itaysk** [aquasecurity/tracee#574] [Development]
Today we expect people to compile the bpf probe on their hosts, either manually using `make bpf` or automatically by letting Tracee do it at start time.
We should be able to pre-compile the probe for different kernel versions, and have Tracee use that.

I'm creating this discussion to discuss some aspect of this feature before we create issues:

## Issue 1 - tracee-ebpf's probe discovery

Today Tracee-ebpf's start sequence is somewhat like:
1. Identify current kernel, and Tracee version, generate expected probe name
2. TRY look for existing probe in `TRACEE_BPF_FILE`
3. TRY look for existing probe next in CWD
4. TRY look for existing probe in `--install-path`
5. TRY building new probe into `--install-path` (according to `--build-policy`)
6. Error

We should add a new step to TRY downloading the probe.
- Step precedence is between 4 and 5 (so that existing probes will be reused)
- The probe should be downloaded into the `--install-path`
- There should be an option to skip this step
- There should be an option to change the remote server.

As with other cases we have experienced, this is a good opportunity to unify the UX:

1. `--bpf build:{never,always,ifneeded}` default: ifneeded (in place of the existing `--build-policy` flag)
2. `--bpf dir:/path/to/dir` default: /tmp/tracee (in place of the existing (`--install-path` flag)
3. `--bpf file:/path/to/bpf.o` (precedes over existing `TRACEE_BPF_FILE` env var)
4. `--bpf remote:{never,always,ifneeded}`
5. `--bpf remote-url:https://url` default: TBD (see Issue 2)

future enhancement (out of scope)
1. User can host their own probe internally
2. Authenticating HTTPS certs
3. Checksum
4. Support more remote protocols (git,s3,etc)

Mentioning here we should use [go-getter](https://github.com/hashicorp/go-getter) in the future which will solve most of those challenges.

## Issue 2 - Build system

We should create an automated build system that compiles the eBPF probe for the different supported kernel version.
- ~~Since we can't cross-compile this needs to be actually spinning up custom VMs, therefore I think managed GitHub Actions is not an option. Perhaps self-hosted GitHub actions could be an option.~~
- we can probably cross-compile by using the correct kernel headers as part of the build. needs POC
- We should be able to update the desired kernel versions matrix independently from Tracee's release lifecycle (for example we decide we need to support a new kernel)
- Preferably able to run docker since it's easier to build the probe this way

## Issue 3 - Probes registry

- We need somewhere to host the compiled artifacts
- Artifacts should be grouped according to Tracee version
- Artifacts should be reference-able by name
- Artifacts should be individually downloadable over HTTP without authentication

Our instinct would be to use GitHub releases, however I don't think it's the best option in this case:
1. Large number of files would complicate up the release page where we send users to find Tracee
2. We may want to add new kernels to the matrix, in which case we will need to modify previous releases. not sure if this feels right (release is a point in time snapshot).
3. Presumably programmatically manipulating the release "Assets" section is not as easy as alternative.
4. It would be nice if we could track download usage

I think we should create our own minimalistic solution based on some cloud storage, and simple proxy.

**Comment by itaysk** [comment]
re the Build system issue, we probably don't need a real kernel (VM) in order to build the probe, we could cross-compile it using just the appropriate headers, as suggested by Amir Jerbi. Needs to be validated but sounds promising

**Comment by grantseltzer** [comment]
Yea if we could have a releases page for tracee-ebpf (or a tracee container) built for various kernel versions that'd be great. I don't feel that we should distribute the tracee go binary separately from the bpf object. When we ship the complete binary we can also embed the bpf object using the go 1.16 embed package (#434) to further simplify things. Taking it a step further we can compile statically to make it even easier.

**Reply by yanivagman** [reply]
Embedding the bpf code in the binary is a good idea! The only problem I can think about with that approach is the size of the created binary if we embed many bpf object files in it. Currently, the size of the bpf.o files is ~1.3MB. Say that we want to support 5 distros, and 10 versions per distro (on average), we get ~65MB extra for the binary. A different approach can be putting the bpf objects as part of the container image.

**Reply by grantseltzer** [reply]
Yea embedding all of them would be a little insane, I pictured only one, and distribute a different binary for each kernel version.

**Comment by itaysk** [comment]
We have a POC of cross-compiling using the target version kernel headers. One thing we need to consider is which kernel headers to use.
1. use each distribution's package packaged headers. 
2. make headers from source

option 2 is easier to implement because it's one code flow, but it potentially is risky, because we don't know how each distro is configured (K configs). 
options 1 is probably safer but is harder to maintain.

@yanivagman what do you think?

**Reply by yanivagman** [reply]
Not sure I understand what is "make headers from source" - do you mean the linux mainline sources?
If so, I don't think that this will work well - in addition to the configuration problem you mentioned, some distros can backport features from newer kernels and change the sources (e.g. redhat).
On the other hand, if you mean downloading the sources from the distros websites instead of using the packages - then in this case I think that the source already includes the .config used, doesn't it?

**Reply by itaysk** [reply]
yes I meant the first one. thanks for confirming

**Reply by itaysk** [reply]
in this case, we need a matrix of `$tracee.$distro.$kernel` right?

**Comment by itaysk** [comment]
@afdesk has been working on this. I think that soon we could open issues/PRs. 
WIP:
https://github.com/falcosecurity/driverkit/issues/100
https://github.com/afdesk/my-driverkit/
https://github.com/afdesk/tracee-building-action
https://github.com/afdesk/tracee-demo-actions
https://github.com/afdesk/tracee/pull/2/files

**Comment by itaysk** [comment]
> in this case, we need a matrix of `$tracee.$distro.$kernel`

We need to better understand what is the kernel version component, meaning what's the granularity we need to build for.
previously we assumed the full version (v.major.minor-patch) but we might not need patch variations as @rafaeldtinoco suggested. @yanivagman WDYT?

**Reply by yanivagman** [reply]
For kprobe programs, the kernel version declared by the bpf code sholuld exactly match the version of the running kernel.
 If I remember correctly, starting from kernel 4.20, this check was removed (but I can't find any reference for that now).
So, I don't think that we actually need to take the patch level into account (because there shuoldn't be any changes between patch variations that might affect the offsets of the kernel structs we are using), just to make sure that the kernel doesn't reject loading these programs as the versions don't match.
Also note that there is a trick to bypass this limitation by declaring a kernel version of 0xFFFFFFFE in the bpf code

**Reply by rafaeldtinoco** [reply]
> previously we assumed the full version (v.major.minor-patch) but we might not need patch variations as @rafaeldtinoco suggested. @yanivagman WDYT?

For some reason I had already in my head that we would be going towards the CO.RE direction with different BTFs because that is something I had previously worked with (and made public). So when I said about not having different data format changes in between distribution kernel update versions I mean that.. for the same BTF, it is VERY likely that we only need a single BTF for major versions.

So, for example, by having CO.RE and 1 BTF file... we can possibly make this single BTF file to work within all 5.1 kernels, then another file to work with all 5.2 kernels, and so on. Data format changes are not very common within stable kernel updates.

Unfortunately we were discussing about multiple BPF object files and I think there was some cross wires there (I had something in my head and you were talking about something else).

> For kprobe programs, the kernel version declared by the bpf code sholuld exactly match the version of the running kernel.

Yes. If there is no CO.RE then there is no relocations to be made and that makes impossible for us to have the same binary in multiple kernels. My proposal is that we have:

BPF CO.RE file + BTFs and, instead of managing multiple BPF objects we would manage much much less BTF ELF files. Example of that is described at [HERE](https://github.com/rafaeldtinoco/portablebpf#portable-libbpf-based-ebpf-code-older-kernels).

> If I remember correctly, starting from kernel 4.20, this check was removed (but I can't find any reference for that now).

Yes, discussion [HERE](https://lore.kernel.org/bpf/B4B873B5-464D-46D7-A0DC-841B08FA34AD@ubuntu.com/) when I found why my portable bpf was failing in 4.15 kernels.

> So, I don't think that we actually need to take the patch level into account (because there shuoldn't be any changes between patch variations that might affect the offsets of the kernel structs we are using), just to make sure that the kernel doesn't reject loading these programs as the versions don't match.

With my suggestion:

Single CO.RE binary + 1 BTF per major release kernel version. Just like as:

https://github.com/rafaeldtinoco/conntracker/tree/main/ebpf/btf

This would make our lives much, much, EASIER. I even tried the same BTF binary in between 5.8 and 4.19 kernels and it works =). When there is a data format change in between distant kernels... libbpf handles the issue by having 0 as the value of the relocated variable (with a warning). That won't be our case if we keep a BTF file per distro major kernel version.

**Note**: We would have to generate BTF files from debug symbols of existent kernels for almost all major releases.. but just a single BTF file per major release version AND for kernels that don't already provide the sysfs BTF file =)... our work decays exponentially...

> Also note that there is a trick to bypass this limitation by declaring a kernel version of 0xFFFFFFFE in the bpf code

With:

```
commit b09a499
Author: Rafael David Tinoco <rafaeldtinoco@ubuntu.com>
Date:   Tue Mar 23 01:09:52 2021

    libbpf: Add bpf object kern_version attribute setter
```

We can get the kernel version during runtime and inform before BPF is loaded (for older kernels).

**Reply by rafaeldtinoco** [reply]
Discussion getting close to:

https://github.com/aquasecurity/tracee/discussions/713

**Reply by itaysk** [reply]
there's no confusion, I think - this discussion is about cross-compiling the probe. IF we go this path, and currently we are, we need to understand the granularity we'll require. The other thread discusses generating BTF which is an alternative approach.

**Reply by rafaeldtinoco** [reply]
> For some reason I had already in my head that we would be going towards the CO.RE direction with different BTFs because that is something I had previously worked with (and made public). So when I said about not having different data format... [snip]

This was related to when we spoke yesterday ^ .

> there's no confusion, I think - this discussion is about cross-compiling the probe. IF we go this path, and currently we are, we need to understand the granularity we'll require. The other thread discusses generating BTF which is an alternative approach.

Got it.

**Reply by yanivagman** [reply]
@rafaeldtinoco I think that one pre-compiled CO-RE enabled bpf code should be enough to support all distros that have BTF support turned on. Eventually, this is what CO-RE comes to solve - porting the same binary to different kernels.
Also see: https://facebookmicrosites.github.io/bpf/blog/2020/02/19/bpf-portability-and-co-re.html#dealing-with-kernel-version-and-configuration-differences

**Reply by rafaeldtinoco** [reply]
> @rafaeldtinoco I think that one pre-compiled CO-RE enabled bpf code should be enough to support all distros that have BTF support turned on. Eventually, this is what CO-RE comes to solve - porting the same binary to different kernels.

Yes, my proposal is:

CO-RE for every kernel that supports BTF. If kernel does not support BTF (because it is old, or it does not have it enabled) we make libbpf to load BTF from elsewhere (an external file) that contains BTF info (that can be generated from DWARF symbols) INSTEAD of having different BPF binaries to each possible environment. 

**This way we would only be CO-RE, no matter what.**

But, like @itaysk said, this is an alternative path right now as the multiple objects approach had already been discussed, is more mature, and is the path chosen right now.

> Also see: https://facebookmicrosites.github.io/bpf/blog/2020/02/19/bpf-portability-and-co-re.html#dealing-with-kernel-version-and-configuration-differences

Yes, I have seen your network changes doing the LINUX_KERNEL_VERSION technique for the bitwise sk_buf fields that turned into full variables. I'm glad they give us the kernel version.

Unfortunately, I think the build time LINUX_KERNEL_VERSION definition (comes from linux-headers) might be wrong in some distros. Debian/Ubuntu are wrong iirc. I know Ubuntu has 4.15.0 as the LINUX_KERNEL_VERSION in its headers... when it should be 4.15.18 in Bionic, for example, the base for their 4.15 kernels.

In those cases we would have to have build tests I guess (for different supported scenarios).

> I discovered that when I introduced the upstream kern_version ebpf attr parameter (I had to specificy 4.15.18 in the eBPF parameter of the object i was loading to 4.15.0 kernels to make it work).

**Reply by yanivagman** [reply]
> Unfortunately, I think the build time LINUX_KERNEL_VERSION definition (comes from linux-headers) might be wrong in some distros. Debian/Ubuntu are wrong iirc. I know Ubuntu has 4.15.0 as the LINUX_KERNEL_VERSION in its headers... when it should be 4.15.18 in Bionic, for example, the base for their 4.15 kernels.

 I know about the kernel version mismatch in ubuntu when you do uname and it is not compatible with what should be given to the bpf version variable. However, this should not be a problem as we use LINUX_KERNEL_VERSION from version.h header, and not taking it from uname.
Anyways, I was referring to CO-RE, where we should not use this macro, but the other methods as described in the link I gave above.

> CO-RE for every kernel that supports BTF. 

ok, so we agree about one CORE bpf object for all kernels with BTF.

> If kernel does not support BTF (because it is old, or it does not have it enabled) we make libbpf to load BTF from elsewhere (an external file) that contains BTF info (that can be generated from DWARF symbols) INSTEAD of having different BPF binaries to each possible environment.
This way we would only be CO-RE, no matter what.

Ok, understood. That's a nice idea!

**Reply by yanivagman** [reply]
Actually, when thinking about your idea, if we have one database that contains BTFs for every kernel we want to support, there is no more dependency in the tracee version itself, and it can be used for every version of tracee, which is awesome. Instead of compiling for every version, we just need to create such BTF database once.
This can actually be a nice idea for a new open source repository, that I'm now asking myself -why no one did it until now? With such db, anyone who want to support ebpf for old kernels with no btf will be able to do this almost instantly

**Reply by rafaeldtinoco** [reply]
> I know about the kernel version mismatch in ubuntu when you do uname and it is not compatible with what should be given to the bpf version variable. However, this should not be a problem as we use LINUX_KERNEL_VERSION from version.h header, and not taking it from uname.

 version.h is wrong as well... and there is no current version variable we can get runtime... (at least I could not find as all version like stuff are macros).

**Reply by rafaeldtinoco** [reply]
> Actually, when thinking about your idea, if we have one database that contains BTFs for every kernel we want to support, there is no more dependency in the tracee version itself, and it can be used for every version of tracee, which is awesome. Instead of compiling for every version, we just need to create such BTF database once.

AND there is a good idea from @itaysk as well. Theoretically our BTF does not need to be complete. We could even strip it to the data we're currently using within tracee only (pahole has the recursive option when displaying data, I wonder if you can format from binary to C and then back to binary). But this other step.

For the BTF database, we could have 1 BTF per major distribution kernel release and supported kernel.

Fedora

Version (Code name) | Release | Kernel
-- | -- | --
27 | 2017-11-14 | 4.13
28 | 2018-05-01 | 4.16
29 | 2018-10-30 | 4.18
30 | 2019-05-07 | 5.0

Ubuntu
Focal (4.19)
Focal HWE (5.4)
Groovy (5.4)
Groovy HWE (5.8)
Next Ubuntu LTS (5.14 ?)
Next Ubuntu LTS HWE (5. ?)

etc...
